File: Backend/WebAPI/Src/Core/Entities/BaseUser.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace Core.Entities;

public abstract class BaseUser
{
    public Guid Id { get; }

    [Column(TypeName = "VARCHAR(64)")]
    public required string UserName { get; set; }

    [Column(TypeName = "VARCHAR(128)")]
    public required string Email { get; set; }

    [Column(TypeName = "VARCHAR(20)")]
    public required string Password { get; set; }

    [Column(TypeName = "VARCHAR(20)")]
    public required string FirstName { get; set; }

    [Column(TypeName = "VARCHAR(100)")]
    public required string LastName { get; set; }

    [Column(TypeName = "VARCHAR(1024)")]
    public string? AvatarURL { get; set; }

    public required DateOnly DateOfBirth { get; set; }
    public required bool IsActive { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public DateTime? LastLogin { get; set; }
}

File: Backend/WebAPI/src/main/java/com/vhl/webapi/entities/specific/BaseUser.java
package com.vhl.webapi.entities.specific;

import com.vhl.webapi.entities.superclass.ICUBaseEntity;
import jakarta.persistence.*;
import lombok.Data;

import java.time.Instant;
import java.time.LocalDate;

@Entity
@Data
@Table(name = "base_users")
@Inheritance(strategy = InheritanceType.JOINED)
public abstract class BaseUser extends ICUBaseEntity {
    @Column(name = "user_name", length = 64, nullable = false, unique = true)
    protected String userName;

    @Column(name = "email", length = 256, nullable = false, unique = true)
    protected String email;

    @Column(name = "password", length = 128, nullable = false)
    protected String password;

    @Column(name = "first_name", length = 50, nullable = false)
    protected String firstName;

    @Column(name = "last_name", length = 100, nullable = false)
    protected String lastName;

    @Column(name = "date_of_birth")
    protected LocalDate dateOfBirth;

    @Column(name = "is_active", nullable = false)
    protected boolean isActive = false;

    @Column(name = "avatar_url", length = 2048)
    protected String avatarUrl;

    @Column(name = "last_login", columnDefinition = "TIMESTAMP")
    protected Instant lastLogin;

    public abstract String getFullRole();
}
